### PR TITLE
Page iterator Implementation

### PR DESCRIPTION
**Pull request by muthurathinam**
# [PageIterator](https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/tasks/PageIteratorTask.md)
For a variety of reasons, collections of entities are often split into pages and each page is returned with a URL to the next page. Sometimes, page granularity provided by the API does not match the requirements of the consumer. This task aims to simplify the life of consumers of paged collections.

**Iterate over all the messages** 
```typescript
async function callingPattern() {
    try {
        // Makes request to fetch mails list. Which is expected to have multiple pages of data.
        let response: PageCollection = await client.api("/me/messages").get();
       // A callback function to be called for every item in the collection. This call back should return boolean indicating whether not to continue the iteration process.
        let callback: PageIteratorCallback = (data) => {
            console.log(data);
            return true;
        };
        // Creating a new page iterator instance with client a graph client instance, page collection response from request and callback
        let pageIterator = new PageIterator(client, response, callback);
        // This iterates the collection until the nextLink is drained out.
        pageIterator.iterate();
    } catch (e) {
        throw e;
    }
}
```

**Stopping and Resuming the iteration**
```typescript
// Populating custom size pages if the api restricts to some maximum size. Lazy loading more data on user prompt or something, stop and resume will do the trick.
async function customSize() {
    try {
        let response: PageCollection = await client.api("/me/messages").get();
        let size = 1000;
        let count = 0;
        let callback: PageIteratorCallback = (data) => {
            console.log(data);
            count++;
            if (count === size) {
                count = 0;
                return false;
            }
            return true;
        };
        let pageIterator = new PageIterator(client, response, callback);
        // This stops iterating over for 1000 entities.
        pageIterator.iterate();

        // Resuming will do start from where it left off and iterate for next 1000 entities.
        // Resume is likely to be called in any user interaction requiring to load more data.
        pageIterator.resume();
    } catch (e) {
        throw e;
    }
}
```

### CODE REVIEW

File: lib/src/tasks/PageIterator.d.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+     * @param {Client} client - The graph client instance
+     * @param {PageCollection} pageCollection - The page collection object
+     * @param {PageIteratorCallback} callBack - The callback function
+     */
+    constructor(client: Client, pageCollection: PageCollection, callback: PageIteratorCallback);
+    /**
+     * Iterates over a collection by enqueuing entries one by one and kicking the callback with the enqueued entry
+     * @return A boolean indicating the continue flag to process next page
+     */
+    iterationHelper(): boolean;
```
[the remark this answers is not in the document]
Comment (muthurathinam) [reply lib/src/tasks/PageIterator.d.ts]: Added in the spec as private, but I missed in the implementation. will change

File: lib/src/tasks/PageIterator.d.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    /**
+     * @async
+     * Iterates over the collection and kicks callback for each item on iteration. Fetches next set of data through nextLink and iterates over again
+     * This happens until the nextLink is drained out or the user responds with a red flag to continue from callback
+     * @return A Promise that resolves to nothing on completion and throws error incase of any discrepancy.
+     */
+    iterate(): Promise<any>;
+    /**
+     * @async
+     * This internally calls the iterate method, Its juts for more readability.
```
[the remark this answers is not in the document]
Comment (muthurathinam) [reply lib/src/tasks/PageIterator.d.ts]: will change

**Comment by muthurathinam** [comment]
@MIchaelMainer Have done the changes that you are requested.

File: src/tasks/PageIterator.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+import { Client } from "../index";
+
+/**
+ * Signature representing PageCollection
+ * @property {any[]} value - The collection value
+ * @property {string} [@odata.nextLink] - The nextLine value
+ * @property {any} Additional - Any number of additional properties (This is to accept the any additional data returned by in the response to the nextLink request)
+ */
+export interface PageCollection {
```
[the remark this answers is not in the document]
Comment (MIchaelMainer) [reply src/tasks/PageIterator.ts]: Or more precisely, we need to know that we can extend this iterate task to access the response headers if we need to in the future.

[the remark this answers is not in the document]
Comment (muthurathinam) [reply src/tasks/PageIterator.ts]: Yeah we can provide such extensibility.

**Pull request merged**